include/zephyr/rtio/rtio.h

```diff
@@ -538,22 +640,58 @@ static inline void rtio_sqe_prep_transceive(struct rtio_sqe *sqe,
  * @param sq_sz Size of the submission queue, must be power of 2
  * @param cq_sz Size of the completion queue, must be power of 2
  */
-#define RTIO_DEFINE(name, exec, sq_sz, cq_sz)							\
-	IF_ENABLED(CONFIG_RTIO_SUBMIT_SEM,							\
-		   (static K_SEM_DEFINE(_submit_sem_##name, 0, K_SEM_MAX_LIMIT)))		\
-	IF_ENABLED(CONFIG_RTIO_CONSUME_SEM,							\
-		   (static K_SEM_DEFINE(_consume_sem_##name, 0, K_SEM_MAX_LIMIT)))		\
-	RTIO_SQ_DEFINE(_sq_##name, sq_sz);							\
-	RTIO_CQ_DEFINE(_cq_##name, cq_sz);							\
-	STRUCT_SECTION_ITERABLE(rtio, name) = {							\
-		.executor = (exec),								\
-		.xcqcnt = ATOMIC_INIT(0),							\
-		IF_ENABLED(CONFIG_RTIO_SUBMIT_SEM, (.submit_sem = &_submit_sem_##name,))	\
-		IF_ENABLED(CONFIG_RTIO_SUBMIT_SEM, (.submit_count = 0,))			\
-		IF_ENABLED(CONFIG_RTIO_CONSUME_SEM, (.consume_sem = &_consume_sem_##name,))	\
-		.sq = (struct rtio_sq *const)&_sq_##name,					\
-		.cq = (struct rtio_cq *const)&_cq_##name,					\
-	};
+/* clang-format off */
+#define RTIO_DEFINE(name, exec, sq_sz, cq_sz)                                                      \
+	_RTIO_DEFINE(name, exec, sq_sz, cq_sz)                                            \
+	}
+/* clang-format on */
+
+/**
+ * @brief Allocate to bss if available
+ *
+ * If CONFIG_USERSPACE is selected, allocate to the rtio_partition bss. Maps to:
+ *   K_APP_BMEM(rtio_partition) static
+ *
+ * If CONFIG_USERSPACE is disabled, allocate as plain static:
+ *   static
+ */
+#define RTIO_BMEM COND_CODE_1(CONFIG_USERSPACE, (K_APP_BMEM(rtio_partition) static), (static))
+
+/**
+ * @brief Allocate as initialized memory if available
+ *
+ * If CONFIG_USERSPACE is selected, allocate to the rtio_partition init. Maps to:
+ *   K_APP_DMEM(rtio_partition) static
+ *
+ * If CONFIG_USERSPACE is disabled, allocate as plain static:
+ *   static
+ */
+#define RTIO_DMEM COND_CODE_1(CONFIG_USERSPACE, (K_APP_DMEM(rtio_partition) static), (static))
+
+/**
+ * @brief Statically define and initialize an RTIO context
+ *
+ * @param name Name of the RTIO
+ * @param exec Symbol for rtio_executor (pointer)
+ * @param sq_sz Size of the submission queue, must be power of 2
+ * @param cq_sz Size of the completion queue, must be power of 2
+ * @param num_blks Number of blocks in the memory pool
+ * @param blk_size The number of bytes in each block
+ * @param balign The block alignment
+ */
+/* clang-format off */
+#define RTIO_DEFINE_WITH_MEMPOOL(name, exec, sq_sz, cq_sz, num_blks, blk_size, balign)             \
+	RTIO_BMEM uint8_t __aligned(WB_UP(balign))                                                 \
+		_mempool_buf_##name[num_blks*WB_UP(blk_size)];	                                   \
+	_SYS_MEM_BLOCKS_DEFINE_WITH_EXT_BUF(_mempool_##name, WB_UP(blk_size), num_blks,		   \
+					    _mempool_buf_##name, RTIO_DMEM);                       \
+	RTIO_DMEM struct rtio_mempool_map_entry _mempool_map_##name[sq_sz] = {0};                  \
```

Comment:
This could be made BMEM (it should be zeroed for you) and avoid the rom space usage.